Stratis.Bitcoin/Utilities/Stopwatch.cs

```diff
@@ -10,78 +10,47 @@ namespace Stratis.Bitcoin.Utilities
+        /// </para>
         /// </summary>
-        private class StopwatchDisposable : IDisposable
-        {
-            /// <summary>Stopwatch to measure elapsed ticks of the code block.</summary>
-            private readonly System.Diagnostics.Stopwatch watch;
+        private readonly Action<long> action;
 
-            /// <summary>
-            /// Action to execute when the measurement is done. 
-            /// <para>
-            /// This is usually a performance counter. 
-            /// The argument of the action is the number of elapsed ticks of the code block.
-            /// </para>
-            /// </summary>
-            private readonly Action<long> action;
-
-            /// <summary>
-            /// Creates a new disposable object and starts the time measurement.
-            /// </summary>
-            /// <param name="watch">Stopwatch to measure elapsed ticks of the code block.</param>
-            /// <param name="action">Action to execute when the measurement is done.</param>
-            public StopwatchDisposable(System.Diagnostics.Stopwatch watch, Action<long> action)
-            {
-                this.watch = watch;
-                this.action = action;
-                watch.Restart();
-            }
-
-            /// <summary>
-            /// Stops the time measurement and calls the action with the measured elapsed ticks.
-            /// </summary>
-            public void Dispose()
-            {
-                this.watch.Stop();
-                this.action(this.watch.Elapsed.Ticks);
-            }
-        }
-
-        /// <summary>Stopwatch to be used by <see cref="StopwatchDisposable"/> instances.</summary>
-        private readonly System.Diagnostics.Stopwatch watch;
-
-        /// <summary>Initializes a new instance of the object.</summary>
-        public Stopwatch()
+        /// <summary>
+        /// Creates a new disposable object and starts the time measurement.
+        /// </summary>
+        /// <param name="action">Action to execute when the measurement is done.</param>
+        public StopwatchDisposable(Action<long> action)
         {
+            this.action = action;
             this.watch = new System.Diagnostics.Stopwatch();
+            this.watch.Start();
         }
 
         /// <summary>
-        /// Starts the time measurement.
+        /// Stops the time measurement and calls the action with the measured elapsed ticks.
         /// </summary>
-        /// <param name="action">Action to execute when the measurement is done. See <see cref="StopwatchDisposable.action"/>.</param>
-        /// <returns>Disposable interface for convenient calling with <c>using</c> keyword.</returns>
-        public IDisposable Start(Action<long> action)
+        public void Dispose()
         {
-            return new StopwatchDisposable(this.watch, action);
+            this.watch.Stop();
+            this.action(this.watch.Elapsed.Ticks);
```

Comment:
should probably check action is not null here, and/or put a Guard on the constructor.
I would say add a null check here so that this class can be used without an action.

Reply:
using this class without an action does not make sense to me - i can't see use case, so i'm happy with the guard

Reply:
if for example someone wants to do:
```
StopwatchDisposable sw;
using(sw = new StopwatchDisposable())
{
// some codez
}
var ticks = sw.ElapsedTicks;

Reply:
OK, so this construction is 5 line construction, which you preferred over
```
Stopwatch sw = Stopwatch.StartNew();
// some code
sw.Stop;
tick = sw.ElapsedTicks;
```
Why on earth would you chose that using here :D 
The only reason why that disposable thing makes sense is that it is 3 line construction (if action is present) and 2 of those lines are brackets, which is possibly better than this 3 line construction I'm having here. But that 5 line thing seems useless to me. Any rationale for that?

Reply:
You're right, but the using also has the advantage of wrapping the code for the user, so that they don't forget to stop the stopwatch. so it has some value, even without an action.

Reply:
I think we can make stop watch more flexible (an override with no action).

Generally I don't even think we need to do it like that, its just confusing I would not use IDisposable at all rather put a stop watch in the context and every measurement take the ElapsedTicks and dispose it at the end.

Reply:
Guys, we are getting off topic I think :-) The purpose of this PR was to get avoid of DTO as an internal measurement engine. Now we are talking about completely turning this upside down. 

I don't think that anyone can actually think of reasonable use case for no action call, therefore I would say - don't implement it until you need it. 

I don't agree that "don't forget to stop the stopwatch" is relevant, because sw.ElapsedTicks would work even without stopping the watch. So you don't actually need to stop it. There is no resource leak if you don't stop it.

Reply:
ok mate, it's fine, I'll check it in.